client/src/extension.ts

```diff
@@ -1,54 +1,66 @@
-const activationByType: {[key in PanelType]: any} = {
-  [PanelType.TOC_EDITOR]: ensureCatch(showTocEditor(PanelType.TOC_EDITOR, resourceRootDir, activePanelsByType)),
-  [PanelType.IMAGE_UPLOAD]: ensureCatch(showImageUpload(PanelType.IMAGE_UPLOAD, resourceRootDir, activePanelsByType)),
-  [PanelType.CNXML_PREVIEW]: ensureCatch(showCnxmlPreview(PanelType.CNXML_PREVIEW, resourceRootDir, activePanelsByType))
-}
-const defaultLocationByType: {[key in PanelType]: vscode.ViewColumn} = {
-  [PanelType.TOC_EDITOR]: vscode.ViewColumn.One,
-  [PanelType.IMAGE_UPLOAD]: vscode.ViewColumn.One,
-  [PanelType.CNXML_PREVIEW]: vscode.ViewColumn.Two
-}
-
-const lazilyFocusOrOpenPanelOfType = (type: PanelType) => {
-  return (...args: any[]) => {
-    if (activePanelsByType[type] != null) {
-      const activePanel = expect(activePanelsByType[type])
-      try {
-        activePanel.reveal(defaultLocationByType[type])
-        return
-      } catch (err) {
-        // Panel was probably disposed
-        return activationByType[type](...args)
-      }
-    }
-    return activationByType[type](...args)
-  }
-}
-
 const extensionExports = {
   activePanelsByType
 }
+let client: LanguageClient
 
-export function activate(context: vscode.ExtensionContext): typeof extensionExports {
+export async function activate(context: vscode.ExtensionContext): Promise<(typeof extensionExports)> {
   // detect Theia. Alert the user if they are running Theia
   expect(process.env.GITPOD_HOST != null && process.env.EDITOR !== 'code' ? undefined : true, 'You seem to be running the Theia editor. Change your Settings in your profile')
 
   client = launchLanguageServer(context)
   populateXsdSchemaFiles(resourceRootDir)
+  await client.onReady()
+
+  const activationByType: {[key in PanelType]: any} = {
+    [PanelType.TOC_EDITOR]: ensureCatch(showTocEditor(PanelType.TOC_EDITOR, resourceRootDir, activePanelsByType, client)),
+    [PanelType.IMAGE_UPLOAD]: ensureCatch(showImageUpload(PanelType.IMAGE_UPLOAD, resourceRootDir, activePanelsByType)),
+    [PanelType.CNXML_PREVIEW]: ensureCatch(showCnxmlPreview(PanelType.CNXML_PREVIEW, resourceRootDir, activePanelsByType))
+  }
+  const defaultLocationByType: {[key in PanelType]: vscode.ViewColumn} = {
+    [PanelType.TOC_EDITOR]: vscode.ViewColumn.One,
+    [PanelType.IMAGE_UPLOAD]: vscode.ViewColumn.One,
+    [PanelType.CNXML_PREVIEW]: vscode.ViewColumn.Two
+  }
+
+  const lazilyFocusOrOpenPanelOfType = (type: PanelType) => {
+    return (...args: any[]) => {
+      if (activePanelsByType[type] != null) {
+        const activePanel = expect(activePanelsByType[type])
+        try {
+          activePanel.reveal(defaultLocationByType[type])
+          return
+        } catch (err) {
+          // Panel was probably disposed
+          return activationByType[type](...args)
+        }
+      }
+      return activationByType[type](...args)
+    }
+  }
+
+  vscode.workspace.onDidChangeWorkspaceFolders(async (event) => {
```

Comment:
This event is fired when you add or remove a folder from your current workspace. On VSCode locally, this event definitely fires, although there are some fallthrough cases, e.g. when a multi-root workspace is first created. Gitpod does seem support this functionality, in case we ever want more than one repo inside a single workspace. That being said, I'm not confident this event will happen often or ever when CMs or vendors use the extension on Gitpod.

This is mostly here so that logic in bundle actions would be a bit less fallible, in that they don't have to go searching for a root workspace all the time. I haven't found any documentation that suggests that `connection.workspace.getWorkspaceFolders()` always orders the root workspace first (this _is_ the case for `vscode.workspace.workspaceFolders`) which means that something like `getRootPathUri` but on the server's side may not be sound.

An alternative would just be to check `connection.workspace.getWorkspaceFolders()`, and throw out a hard error when there is more than one entry and rely on the fact that someone making a multi-root workspace is a very rare thing (and way more likely to be done by a dev than a content editor)

Thoughts?

Reply:
Oh, I think it's fine to leave it as is -- Mostly I was wondering if there's something about workspaces that I wasn't understanding correctly. The ☝️  answered my questions, though. Thanks for the explanation!